### ISSUE TITLE
ScalaJack 6.0.x is one of the slowest JSON parsers

### ISSUE THREAD

**Issue by plokhotnyuk** [gzoller/ScalaJack#184]
![image](https://user-images.githubusercontent.com/890289/65895969-937f8a00-e3ac-11e9-9664-29619a21cbbf.png)
https://github.com/plokhotnyuk/jsoniter-scala/pull/402

**Comment by gzoller** [comment]
It depends!  Closing this ticket.  Explanation on the DOS ticket.

**Comment by plokhotnyuk** [comment]
@gzoller would you be interested in adding ScalaJack tests for [a set of 100+ JMH-based benchmarks](https://github.com/plokhotnyuk/jsoniter-scala/tree/master/jsoniter-scala-benchmark/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/benchmark)?

It can provide you not only throughout and allocation scores, but 1st of all it will give you tools for checking what is happening under hood for different data types and kinds of JSON messages.

**Comment by gzoller** [comment]
I would indeed be interested, but first let me make my change so users have the option of power or speed, i.e. they can prefetch the TypeAdapter and avoid repetative tt.tpe calls.

SJ 6.0.5 should be cut in the next few days :-)

**Comment by gzoller** [comment]
Hello!  Sorry for the delay--haven't forgotten!  I'm still very interested in your offer but now I've decided to do a bare-metal reassembly of ScalaJack, making some key re-factorings.  It's gonna take some time.  

I did an initial experiment, code named Sinbad, to see if my ideas will work better and the results are promising, so I'd like to complete the work.  Reworking all the tests is gonna take a while.  I did discover my "path" indicator in the error output is not cheap to maintain.  I was sad to cut it out, but it did boost speed.  (I wish the JVM didn't throw away the parameters to the methods in the call stack on Exception!)

I'm comparing SJ5, SJ6 (which did fall down noticeably in performance), Circe, and 2 flavors of Sinbad.  Among the number of changes, I'm skipping JSON tokenization and parsing directly to finished product, which does save cycles.  I'm about to re-install the trait-handling code so I'll see what type hint scan-ahead costs, and compare cost of alternatives.

```
[info] Benchmark                       Mode  Cnt      Score      Error  Units
[info] BaseBenchmarks.read5           thrpt   10  17100.067 ± 2333.296  ops/s
[info] BaseBenchmarks.read6           thrpt   10  11417.863 ±  433.097  ops/s
[info] BaseBenchmarks.readCirce       thrpt   10  19814.908 ±  236.127  ops/s
[info] BaseBenchmarks.readSinbad      thrpt   10  21201.970 ±  556.440  ops/s
[info] BaseBenchmarks.readSinbadFast  thrpt   10  26406.982 ±  311.447  ops/s
```
